DDOT_module/parent_classes/marginals_analyzer_base.py

```diff
@@ -0,0 +1,297 @@
+        We use convention that such label if of the form "q5q8q12..." etc., hence it is bitsring of
+        qubits labels starting from "q".
+        -marginal_probability_vector marginal distribution stored as vector
+
+        NOTE: when user does not provide marginals_dictionary we create it during class initialization.
+        To this aim, we create "key_dependent_dicts"
+
+
+        """
+
+        if marginals_dictionary is None:
+            # If user does not provide dictionary with marginals, we create template.
+            # Instead of standard dictionaries, we use ones that are "key dependent" (see description
+            # of that function), which is useful for some calculations.
+            marginals_dictionary = {key: fda.KeyDependentDictForMarginals()
+                                    for key in results_dictionary.keys()}
+
+        # set initial values of class properties
+        self._results_dictionary = results_dictionary
+        self._marginals_dictionary = marginals_dictionary
+        self._bitstrings_right_to_left = bitstrings_right_to_left
+
+    @property
+    def results_dictionary(self) -> Dict[str, Dict[str, int]]:
+        return self._results_dictionary
+
+    @results_dictionary.setter
+    def results_dictionary(self, results_dictionary: Dict[str, Dict[str, int]]) -> None:
+        self._results_dictionary = results_dictionary
+
+    @property
+    def marginals_dictionary(self) -> Dict[str, Dict[str, np.ndarray]]:
+        return self._marginals_dictionary
+
+    @marginals_dictionary.setter
+    def marginals_dictionary(self, marginals_dictionary: Dict[str, Dict[str, np.ndarray]]) -> None:
+        self._marginals_dictionary = marginals_dictionary
+
+    @staticmethod
+    def get_qubits_key(list_of_qubits: List[int]) -> str:
+        """ from subset of qubit indices get the string that labels this subset
+            using convention 'q5q6q12...' etc.
+        :param list_of_qubits: labels of qubits
+
+        :return: string label for qubits
+
+         NOTE: this function is "dual" to self.get_qubit_indices_from_string.
+        """
+
+        return 'q' + 'q'.join([str(s) for s in list_of_qubits])
+
+    @staticmethod
+    def get_qubit_indices_from_string(qubits_string: str,
+                                      with_q: Optional[bool] = False):
+
+        """Return list of qubit indices from the string of the form "q0q1q22q31"
+        :param qubits_string: string which has the form of "q" followed by qubit index
+        :param with_q: specify whether returned indices should be in form of string with letter
+
+        :return: list of qubit indices:
+
+        depending on value of parameter "with_q" the mapping will be one of the following:
+        if with_q:
+            'q1q5q13' -> ['q1','q5','q13']
+        else:
+            'q1q5q13' -> [1,5,13]
+
+        NOTE: this function is "dual" to self.get_qubits_key.
+        """
+
+        numbers = re.findall(r'\d+', qubits_string)
```

Comment:
```suggestion
        additional_char = ''
    if with_q:
        additional_char = 'q'
    return [f"{additional_char}{qubit_index}" for qubit_index in qubits_string.split('q') if len(qubit_index) > 0]
```
Here's a re independent approach. I'm not sure if split is faster though, so I'm just leaving it as a food for thought.